File: src/main/java/com/kisman/cc/module/render/Trails.java
package com.kisman.cc.module.render;

import com.kisman.cc.Kisman;
import com.kisman.cc.event.events.EventPlayerMotionUpdate;
import com.kisman.cc.module.Category;
import com.kisman.cc.module.Module;
import com.kisman.cc.settings.Setting;
import com.kisman.cc.util.RenderUtil;
import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityExpBottle;
import net.minecraft.entity.projectile.EntityArrow;
import net.minecraft.entity.projectile.EntityThrowable;
import net.minecraft.util.math.Vec3d;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import org.lwjgl.opengl.GL11;

import java.awt.*;
import java.util.*;
import java.util.List;

public class Trails extends Module {
    private Setting lineWidth = new Setting("LineWidth", this, Double.longBitsToDouble(Double.doubleToLongBits(17.27955007873751) ^ 0x7FC14790980DC597L), Double.longBitsToDouble(Double.doubleToLongBits(93.95772593779462) ^ 0x7FB77D4B61BB56F7L), Double.longBitsToDouble(Double.doubleToLongBits(0.1654465991727702) ^ 0x7FD12D5AAA573A5DL), false);
    private Setting color = new Setting("Color", this, "Color", new float[] {1, 1, 1, 1});

    public Color color1;
    public Map<Entity, List<Vec3d>> renderMap = new HashMap<Entity, List<Vec3d>>();

    public Trails() {
        super("Trails", "", Category.RENDER);

        setmgr.rSetting(lineWidth);
        setmgr.rSetting(color);
    }

    public void onEnable() {
        Kisman.EVENT_BUS.subscribe(listener);
    }

    public void onDisable() {
        Kisman.EVENT_BUS.unsubscribe(listener);
    }

    @SubscribeEvent
    public void onRenderWorld(RenderWorldLastEvent event) {
        if (mc.player == null || mc.world == null) {
            return;
        }
        for (Entity entity : mc.world.loadedEntityList) {
            int[] counter = new int[]{1};
            if (!this.renderMap.containsKey(entity)) continue;
            GlStateManager.pushMatrix();
            RenderUtil.GLPre((float) lineWidth.getValDouble());
            GlStateManager.enableBlend();
            GlStateManager.disableTexture2D();
            GlStateManager.depthMask((boolean)false);
            GlStateManager.disableDepth();
            GlStateManager.tryBlendFuncSeparate((GlStateManager.SourceFactor)GlStateManager.SourceFactor.SRC_ALPHA, (GlStateManager.DestFactor)GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, (GlStateManager.SourceFactor)GlStateManager.SourceFactor.ONE, (GlStateManager.DestFactor)GlStateManager.DestFactor.ZERO);
            GL11.glLineWidth((float) lineWidth.getValDouble());
            GL11.glBegin((int)1);
            for (int i = 0; i < this.renderMap.get(entity).size() - 1; ++i) {
                GL11.glColor4f(this.color.getR() / 255, this.color.getG() / 255, (float)((float)this.color.getB() / 255), 1);
                Vec3d pos = updateToCamera(this.renderMap.get(entity).get(i));
                Vec3d pos2 = updateToCamera(this.renderMap.get(entity).get(i + 1));
                GL11.glVertex3d((double)pos.x, (double)pos.y, (double)pos.z);
                GL11.glVertex3d((double)pos2.x, (double)pos2.y, (double)pos2.z);
                counter[0] = counter[0] + 1;
            }
            GL11.glEnd();
            GlStateManager.resetColor();
            GlStateManager.enableDepth();
            GlStateManager.depthMask((boolean)true);
            GlStateManager.enableTexture2D();
            GlStateManager.disableBlend();
            RenderUtil.GlPost();
            GlStateManager.popMatrix();
        }
    }

    @EventHandler
    private final Listener<EventPlayerMotionUpdate> listener = new Listener<>(event -> {
        if (mc.player == null || mc.world == null) {
            return;
        }
        for (Entity entity : mc.world.loadedEntityList) {
            if (!(entity instanceof EntityThrowable) && !(entity instanceof EntityArrow) || entity instanceof EntityExpBottle) continue;
            ArrayList<Vec3d> vectors = this.renderMap.get(entity) != null ? (ArrayList<Vec3d>) this.renderMap.get(entity) : new ArrayList<Vec3d>();
            vectors.add(new Vec3d(entity.posX, entity.posY, entity.posZ));
            this.renderMap.put(entity, vectors);
        }
    });

    public static Vec3d updateToCamera(Vec3d vec) {
        return new Vec3d(vec.x - mc.getRenderManager().viewerPosX, vec.y - mc.getRenderManager().viewerPosY, vec.z - mc.getRenderManager().viewerPosZ);
    }
}

File: src/main/java/com/kisman/cc/features/module/render/Trails.kt
package com.kisman.cc.features.module.render

import com.kisman.cc.features.module.Category
import com.kisman.cc.features.module.Module
import com.kisman.cc.settings.Setting
import com.kisman.cc.settings.types.number.NumberType
import com.kisman.cc.util.Colour
import com.kisman.cc.util.TimerUtils
import com.kisman.cc.util.render.Rendering
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec3d
import net.minecraft.world.DimensionType
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import org.lwjgl.opengl.GL11

/**
 * @author _kisman_
 * @since 18:25 of 06.06.2022
 */
class Trails : Module(
    "Trails",
    "nice",
    Category.RENDER
) {
    private val color = register(Setting("Color", this, Colour(255, 255, 255, 255)))
    private val delay = register(Setting("Delay", this, 10.0, 1.0, 1000.0, NumberType.TIME))
    private val fade = register(Setting("Fade", this, true))
    private val fadeDelay = register(Setting("Fade Delay", this, 500.0, 10.0, 10000.0, NumberType.TIME).setVisible { fade.valBoolean })
    private val lineWidth = register(Setting("Line Width", this, 1.0, 0.1, 5.0, false))

    private val timer = TimerUtils()

    private val positions = ArrayList<Trace>()
    private var trace : Trace? = null

    companion object {
        val origin = Vec3d(8.0, 64.0, 8.0)
    }

    override fun onEnable() {
        super.onEnable()
        timer.reset()
        positions.clear()
        trace = null
    }

    @SubscribeEvent fun onRenderWorld(event : RenderWorldLastEvent) {
        if(timer.passedMillis(delay.valLong)) {
            val vec = mc.player.positionVector

            if(trace == null) {
                trace = Trace(
                    0,
                    null,
                    mc.world.provider.dimensionType,
                    vec,
                    ArrayList()
                )
            }

            val trace1 = trace!!.trace
            val vec3d = if(trace1.isEmpty()) vec else trace1[trace1.size - 1].pos

            if(trace1.isNotEmpty()) {
                positions.add(trace!!)
                trace = Trace(
                    positions.size + 1,
                    null,
                    mc.world.provider.dimensionType,
                    vec,
                    trace1
                )
            }

            if(trace1.isEmpty() || !vec.equals(vec3d)) {
                trace1.add(TracePos(
                    vec,
                    System.currentTimeMillis() + fadeDelay.valInt
                ))
            }

            timer.reset()
        }

        if(trace != null) {
            Rendering.start()
            GL11.glLineWidth(lineWidth.valFloat)

            for(pos in positions) {
                GL11.glBegin(GL11.GL_LINE_STRIP)
                for(tr in pos.trace) {
                    render(tr)
                }
                GL11.glEnd()

                if(fade.valBoolean) {
                    pos.trace.removeIf {
                        it.shouldBeRemove()
                    }
                }
            }

            GL11.glBegin(GL11.GL_LINE_STRIP)

            for(tr in trace?.trace!!) {
                render(tr)
            }

            if(fade.valBoolean) {
                trace?.trace!!.removeIf {
                    it.shouldBeRemove()
                }
            }

            GL11.glEnd()

            Rendering.end()
        }
    }

    private fun render(trace : TracePos) {
        val alpha = if(fade.valBoolean) {
            MathHelper.clamp((color.colour.a1 / fadeDelay.valInt) * (trace.time - System.currentTimeMillis()), 0f, color.colour.b1) / color.colour.b1
        } else {
            color.colour.a1
        }
        GL11.glColor4f(
            color.colour.r1,
            color.colour.g1,
            color.colour.b1,
            alpha
        )

        GL11.glVertex3d(
            trace.pos.x - mc.renderManager.renderPosX,
            trace.pos.y - mc.renderManager.renderPosY,
            trace.pos.z - mc.renderManager.renderPosZ
        )
    }

    class Trace(
        var index: Int,
        var name: String?,
        var type: DimensionType,
        var pos: Vec3d,
        val trace: ArrayList<TracePos>
    )

    class TracePos(
        var pos : Vec3d,
        var time : Long
    ) {
        val timer = TimerUtils()

        fun shouldBeRemove() : Boolean {
            return timer.passedMillis(2000L)
        }
    }
}